File: python/tests/test_rvo.py
"""RVO (van den Berg, Lin & Manocha 2008): VO with the cone apex shifted toward
the midpoint of both agents' velocities so a symmetric encounter doesn't
oscillate. Verified the same way as VO: reciprocal avoidance on the shared
multi-body scenarios, and an honest failure when there is no room to avoid."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import numpy as np
import pytest
import yaml
from conftest import REPO_ROOT

from navigation.core.params import ParamError, ParamSet
from navigation.core.types import RobotState
from navigation.local_planning.simulation import SimConfig, SimStatus
from navigation.local_planning.velocity.agent_scenario import load_agent_scenario
from navigation.local_planning.velocity.agent_sim import AgentSpec, simulate_agents
from navigation.local_planning.velocity.rvo import Rvo
from navigation.maps.loader import load_map
from navigation.maps.occupancy_grid import OccupancyGrid2D

_ALGO = "rvo"
_CONFIG_PATH = REPO_ROOT / "configs" / "local_planning" / f"{_ALGO}.yaml"
_MAPS_DIR = REPO_ROOT / "maps"


def _config(tmp_path: Path, **overrides: Any) -> ParamSet:
    doc = yaml.safe_load(_CONFIG_PATH.read_text())
    for entry in doc["params"]:
        if entry["name"] in overrides:
            entry["default"] = overrides[entry["name"]]
    out = tmp_path / f"{_ALGO}.yaml"
    out.write_text(yaml.safe_dump(doc), encoding="utf-8")
    return ParamSet.from_yaml(out)


def _sim_config(params: ParamSet) -> SimConfig:
    return SimConfig(
        control_dt=params.get_float("control_dt"),
        max_steps=params.get_int("max_steps"),
        goal_tolerance=params.get_float("goal_tolerance"),
        footprint_radius=params.get_float("footprint_radius"),
        stall_window=params.get_int("stall_window"),
        stall_distance=params.get_float("stall_distance"),
    )


def _run_scenario(scenario_name: str, params: ParamSet) -> list[Any]:
    scenario = load_agent_scenario(_MAPS_DIR / "scenarios" / "velocity" / f"{scenario_name}.yaml")
    grid = load_map(scenario.map_path)
    planners = [Rvo(params) for _ in scenario.agents]
    return simulate_agents(planners, list(scenario.agents), grid, _sim_config(params))


# --- (a) reciprocal avoidance succeeds on the shared multi-body scenarios -----
def test_head_on_two_agents_avoid_and_reach(tmp_path: Path) -> None:
    params = _config(tmp_path)
    results = _run_scenario("head_on", params)
    assert all(r.status != SimStatus.COLLISION for r in results)
    assert all(r.min_pair_clearance >= 0.0 for r in results)
    assert all(r.status == SimStatus.REACHED for r in results)


def test_circle_swap_four_agents_avoid_and_reach(tmp_path: Path) -> None:
    params = _config(tmp_path)
    results = _run_scenario("circle_swap", params)
    assert all(r.status != SimStatus.COLLISION for r in results)
    assert all(r.min_pair_clearance >= 0.0 for r in results)
    assert all(r.status == SimStatus.REACHED for r in results)


# --- (b) an honest failure: no room to avoid a head-on scripted mover ---------
def test_narrow_corridor_head_on_mover_fails_honestly(tmp_path: Path) -> None:
    params = _config(tmp_path, max_steps=300)
    resolution = 0.1
    band_rows = 8  # 0.8 m: below the 1.2 m two 0.3 m-radius bodies need to pass
    margin_rows = 10
    cols = 100
    pixels = np.zeros((band_rows + 2 * margin_rows, cols), dtype=np.uint16)
    pixels[margin_rows : margin_rows + band_rows, :] = 255
    grid = OccupancyGrid2D(pixels=pixels, resolution=resolution, origin=(0.0, 0.0, 0.0))
    corridor_y = grid.cell_to_world(margin_rows + band_rows // 2, 0)[1]

    planner_spec = AgentSpec(
        start=RobotState(pose=(1.0, corridor_y, 0.0)), goal=(9.0, corridor_y, 0.0), radius=0.3
    )
    mover_spec = AgentSpec(
        start=RobotState(pose=(9.0, corridor_y, 0.0)),
        goal=(1.0, corridor_y, 0.0),
        radius=0.3,
        scripted_velocity=(-1.0, 0.0),
    )
    results = simulate_agents(
        [Rvo(params), None], [planner_spec, mover_spec], grid, _sim_config(params)
    )
    assert results[0].status in (SimStatus.COLLISION, SimStatus.STALLED)
    assert results[0].status != SimStatus.REACHED


# --- (c) parameter validation failures at load time ---------------------------
def test_time_horizon_at_zero_rejected_at_load_time(tmp_path: Path) -> None:
    with pytest.raises(ParamError):
        _config(tmp_path, time_horizon=0.0)


def test_speed_samples_zero_rejected_at_load_time(tmp_path: Path) -> None:
    with pytest.raises(ParamError):
        _config(tmp_path, speed_samples=0)


# --- (d) determinism: identical rerun -> bit-identical per-agent trajectories -
def test_rerun_is_deterministic(tmp_path: Path) -> None:
    params = _config(tmp_path)
    scenario = load_agent_scenario(_MAPS_DIR / "scenarios" / "velocity" / "circle_swap.yaml")
    grid = load_map(scenario.map_path)
    config = _sim_config(params)

    first = simulate_agents(
        [Rvo(params) for _ in scenario.agents], list(scenario.agents), grid, config
    )
    second = simulate_agents(
        [Rvo(params) for _ in scenario.agents], list(scenario.agents), grid, config
    )

    for r1, r2 in zip(first, second, strict=True):
        assert r2.status == r1.status
        assert r2.steps == r1.steps
        assert len(r2.trajectory) == len(r1.trajectory)
        for p1, p2 in zip(r1.trajectory, r2.trajectory, strict=True):
            assert p1 == pytest.approx(p2)


# --- reciprocity=0 collapses onto plain VO (behavioral contract, not an echo) -
def test_reciprocity_zero_matches_plain_vo_apex() -> None:
    from navigation.local_planning.velocity._velocity_obstacle import rvo_apex

    v_self = (1.0, 0.5)
    v_other = (-0.5, 0.2)
    assert rvo_apex(v_self, v_other, 0.0) == pytest.approx(v_other)
    assert rvo_apex(v_self, v_other, 1.0) == pytest.approx(v_self)

File: cpp/tests/test_rvo.cpp
#include <fstream>
#include <optional>
#include <sstream>
#include <string>
#include <vector>

#include <gtest/gtest.h>

#include "navigation/core/params.hpp"
#include "navigation/core/types.hpp"
#include "navigation/local_planning/simulation.hpp"
#include "navigation/local_planning/velocity/agent_scenario.hpp"
#include "navigation/local_planning/velocity/agent_sim.hpp"
#include "navigation/local_planning/velocity/rvo.hpp"
#include "navigation/local_planning/velocity/velocity_obstacle.hpp"
#include "navigation/maps/loader.hpp"
#include "navigation/maps/occupancy_grid.hpp"
#include "test_util.hpp"

// RVO (van den Berg, Lin & Manocha 2008): VO with the cone apex shifted
// toward the midpoint of both agents' velocities so a symmetric encounter
// doesn't oscillate. Verified the same way as VO: reciprocal avoidance on the
// shared multi-body scenarios, and an honest failure when there is no room
// to avoid.

using namespace navigation;
using core::ParamSet;
using core::Point;
using core::Pose;
using core::RobotState;
using local_planning::AgentResult;
using local_planning::AgentSpec;
using local_planning::load_agent_scenario;
using local_planning::rvo_apex;
using local_planning::simulate_agents;
using local_planning::Rvo;
using local_planning::SimConfig;
using local_planning::SimStatus;
using local_planning::VelocityObstaclePlanner;

namespace {

std::string config_path() { return test::repo_path("configs/local_planning/rvo.yaml"); }

std::string read_file(const std::string& path) {
  std::ifstream in(path);
  std::ostringstream ss;
  ss << in.rdbuf();
  return ss.str();
}

ParamSet config_with_override(const std::string& param_name, const std::string& new_value) {
  std::string text = read_file(config_path());
  std::string needle = "name: " + param_name + "\n";
  size_t pos = text.find(needle);
  if (pos == std::string::npos) throw std::runtime_error("test: param not found: " + param_name);
  size_t default_pos = text.find("default:", pos);
  size_t line_end = text.find('\n', default_pos);
  text = text.substr(0, default_pos) + "default: " + new_value + text.substr(line_end);
  return ParamSet::from_yaml(test::write_temp("rvo_override.yaml", text));
}

SimConfig sim_config_from(const ParamSet& params) {
  return SimConfig{params.get_float("control_dt"),     params.get_int("max_steps"),
                   params.get_float("goal_tolerance"), params.get_float("footprint_radius"),
                   params.get_int("stall_window"),     params.get_float("stall_distance")};
}

maps::OccupancyGrid2D& as_grid(core::MapBase& m) { return *dynamic_cast<maps::OccupancyGrid2D*>(&m); }

std::vector<AgentResult> run_scenario(const std::string& scenario_name, const ParamSet& params) {
  auto scenario =
      load_agent_scenario(test::repo_path("maps/scenarios/velocity/" + scenario_name + ".yaml"));
  auto map = maps::load_map(scenario.map_path);
  auto& grid = as_grid(*map);
  std::vector<Rvo> planners;
  planners.reserve(scenario.agents.size());
  for (size_t i = 0; i < scenario.agents.size(); ++i) planners.emplace_back(params);
  std::vector<VelocityObstaclePlanner*> planner_ptrs;
  planner_ptrs.reserve(planners.size());
  for (Rvo& p : planners) planner_ptrs.push_back(&p);
  return simulate_agents(planner_ptrs, scenario.agents, grid, sim_config_from(params));
}

}  // namespace

// --- (a) reciprocal avoidance succeeds on the shared multi-body scenarios ----
TEST(Rvo, HeadOnTwoAgentsAvoidAndReach) {
  auto params = ParamSet::from_yaml(config_path());
  auto results = run_scenario("head_on", params);
  for (const AgentResult& r : results) {
    EXPECT_NE(r.status, SimStatus::COLLISION);
    EXPECT_GE(r.min_pair_clearance, 0.0);
    EXPECT_EQ(r.status, SimStatus::REACHED);
  }
}

TEST(Rvo, CircleSwapFourAgentsAvoidAndReach) {
  auto params = ParamSet::from_yaml(config_path());
  auto results = run_scenario("circle_swap", params);
  for (const AgentResult& r : results) {
    EXPECT_NE(r.status, SimStatus::COLLISION);
    EXPECT_GE(r.min_pair_clearance, 0.0);
    EXPECT_EQ(r.status, SimStatus::REACHED);
  }
}

// --- (b) an honest failure: no room to avoid a head-on scripted mover --------
TEST(Rvo, NarrowCorridorHeadOnMoverFailsHonestly) {
  auto params = config_with_override("max_steps", "300");
  const double resolution = 0.1;
  const int band_rows = 8;  // 0.8 m: below the 1.2 m two 0.3 m-radius bodies need to pass
  const int margin_rows = 10;
  const int cols = 100;
  std::vector<bool> free_cells(static_cast<size_t>(band_rows + 2 * margin_rows) * cols, false);
  for (int r = margin_rows; r < margin_rows + band_rows; ++r) {
    for (int c = 0; c < cols; ++c) free_cells[static_cast<size_t>(r) * cols + c] = true;
  }
  maps::OccupancyGrid2D grid(band_rows + 2 * margin_rows, cols, resolution, 0.0, 0.0,
                             std::move(free_cells));
  double corridor_y =
      grid.cell_to_world(core::Cell{margin_rows + band_rows / 2, 0}).y;

  AgentSpec planner_spec{RobotState{Pose{1.0, corridor_y, 0.0}, 0.0, 0.0},
                        Pose{9.0, corridor_y, 0.0}, 0.3, std::nullopt};
  AgentSpec mover_spec{RobotState{Pose{9.0, corridor_y, 0.0}, 0.0, 0.0}, Pose{1.0, corridor_y, 0.0},
                      0.3, Point{-1.0, 0.0}};

  Rvo planner(params);
  std::vector<VelocityObstaclePlanner*> planners{&planner, nullptr};
  std::vector<AgentSpec> specs{planner_spec, mover_spec};
  auto results = simulate_agents(planners, specs, grid, sim_config_from(params));

  EXPECT_TRUE(results[0].status == SimStatus::COLLISION || results[0].status == SimStatus::STALLED);
  EXPECT_NE(results[0].status, SimStatus::REACHED);
}

// --- (c) parameter validation failures at load time ---------------------------
TEST(Rvo, TimeHorizonAtZeroRejectedAtLoadTime) {
  std::string bad = test::write_temp(
      "rvo_time_horizon.yaml",
      "algorithm: rvo\ncategory: local_planning\nparams:\n"
      "  - {name: time_horizon, type: float, default: 0.0, min: 0.05, max: 20.0, description: at zero}\n");
  EXPECT_THROW(ParamSet::from_yaml(bad), std::runtime_error);
}

TEST(Rvo, SpeedSamplesZeroRejectedAtLoadTime) {
  std::string bad = test::write_temp(
      "rvo_speed_samples.yaml",
      "algorithm: rvo\ncategory: local_planning\nparams:\n"
      "  - {name: speed_samples, type: int, default: 0, min: 1, max: 50, description: at zero}\n");
  EXPECT_THROW(ParamSet::from_yaml(bad), std::runtime_error);
}

// --- (d) determinism: identical rerun -> bit-identical per-agent trajectories -
TEST(Rvo, RerunIsDeterministic) {
  auto params = ParamSet::from_yaml(config_path());
  auto scenario = load_agent_scenario(test::repo_path("maps/scenarios/velocity/circle_swap.yaml"));
  auto map = maps::load_map(scenario.map_path);
  auto& grid = as_grid(*map);
  SimConfig config = sim_config_from(params);

  auto run_once = [&]() {
    std::vector<Rvo> storage;
    storage.reserve(scenario.agents.size());
    for (size_t i = 0; i < scenario.agents.size(); ++i) storage.emplace_back(params);
    std::vector<VelocityObstaclePlanner*> ptrs;
    ptrs.reserve(storage.size());
    for (Rvo& p : storage) ptrs.push_back(&p);
    return simulate_agents(ptrs, scenario.agents, grid, config);
  };

  auto first = run_once();
  auto second = run_once();

  ASSERT_EQ(second.size(), first.size());
  for (size_t i = 0; i < first.size(); ++i) {
    EXPECT_EQ(second[i].status, first[i].status);
    EXPECT_EQ(second[i].steps, first[i].steps);
    ASSERT_EQ(second[i].trajectory.size(), first[i].trajectory.size());
    for (size_t j = 0; j < first[i].trajectory.size(); ++j) {
      EXPECT_NEAR(first[i].trajectory[j].x, second[i].trajectory[j].x, 1e-9);
      EXPECT_NEAR(first[i].trajectory[j].y, second[i].trajectory[j].y, 1e-9);
      EXPECT_NEAR(first[i].trajectory[j].theta, second[i].trajectory[j].theta, 1e-9);
    }
  }
}

// --- reciprocity=0 collapses onto plain VO (behavioral contract, not an echo) -
TEST(Rvo, ReciprocityZeroMatchesPlainVoApex) {
  core::Point v_self{1.0, 0.5};
  core::Point v_other{-0.5, 0.2};
  core::Point at_zero = rvo_apex(v_self, v_other, 0.0);
  core::Point at_one = rvo_apex(v_self, v_other, 1.0);
  EXPECT_NEAR(at_zero.x, v_other.x, 1e-9);
  EXPECT_NEAR(at_zero.y, v_other.y, 1e-9);
  EXPECT_NEAR(at_one.x, v_self.x, 1e-9);
  EXPECT_NEAR(at_one.y, v_self.y, 1e-9);
}
